go-controller/pkg/ovn/policy.go

```diff
@@ -449,31 +442,37 @@ func (oc *Controller) updateACLLoggingForPolicy(np *networkPolicy, logLevel stri
 	if !np.created {
 		return NetworkPolicyNotCreated
 	}
-
-	acls := oc.buildNetworkPolicyACLs(np, logLevel)
-	ops, err := libovsdbops.UpdateACLsLoggingOps(oc.nbClient, nil, acls...)
-	if err != nil {
-		return err
+	// Predicate for given network policy ACLs
+	p := func(item *nbdb.ACL) bool {
+		return item.ExternalIDs[namespaceACLExtIdKey] == np.namespace && item.ExternalIDs[policyACLExtIdKey] == np.name
 	}
-	_, err = libovsdbops.TransactAndCheck(oc.nbClient, ops)
-	return err
+	return UpdateACLLoggingWithPredicate(oc.nbClient, p, aclLogging)
 }
 
 func (oc *Controller) setNetworkPolicyACLLoggingForNamespace(ns string, nsInfo *namespaceInfo) error {
-	var ovsDBOps []ovsdb.Operation
-	for _, policyType := range []knet.PolicyType{knet.PolicyTypeIngress, knet.PolicyTypeEgress} {
-		denyACL, _ := buildDenyACLs(ns, "", targetPortGroupName(nsInfo.portGroupIngressDenyName, nsInfo.portGroupEgressDenyName, policyType), nsInfo.aclLogging.Deny, policyType)
-		var err error
-		ovsDBOps, err = libovsdbops.UpdateACLsLoggingOps(oc.nbClient, ovsDBOps, denyACL)
-		if err != nil {
-			return err
-		}
+	// update namespace-scoped default deny ACLs
+
+	// default network policy ACLs don't have any special name or externalIDs,
+	// therefore we can get all ACLs referenced by defaultDeny port groups
+	defaultDenyPGPred := func(item *nbdb.PortGroup) bool {
```

Comment:
TBH, that's the only part where I got lost a bit; I can't judge if the old code and the new code do the same thing, but I trust that we've got enough unit tests and that you tested this. The new code looks sound, though :-)

Reply:
Let me still try to explain - even though I probably can't add any code comments to make it more clear.
Before it didn't try to find existing ACLs in the db, but build them from scratch and then match by (match, action, priority, direction). So it was executing the same code that actually creates there ACLs here https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/policy.go#L374-L377 which is error-prone (if we change match at some point, already existing ACLs with old match won't be updated)

And since there is really no sane way to find these acls by their fields, because they don't have any special ExternalIDs or name, we use the fact that default Deny PG should be referencing these ACL. So we find default port groups first, then we find referenced ACLs and only choose the ones with "deny" action because here https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/policy.go#L362 you can see that allow ACLs have acl logging set to "" (meaning we explicitly ignore it for logging).

I know that is complicated :) but hopefully it is a little more clear now.

Reply:
Not sure this was error prone at all but what is being proposed has potential for better scale. But

It just works much better for scale if you get the ACLs for the PGs and then you filter those ACLs yourself instead of using libovsdb predicates. That way instead of iterating through the full table of ACLs you just check the handful contained in the PGs.